File: Tree/src/main/java/Solution208.java
import java.util.HashMap;

class Trie {

    class Node {
        HashMap<Character, Node> child;
        int count;

        public Node() {
            count = 0;
            child = new HashMap<>();
        }
    }

    Node root;

    public Trie() {
        this.root = new Node();
    }

    public void insert(String word) {
        find(word, true, false);
    }

    public boolean search(String word) {
        return find(word, false, false);
    }

    public boolean startsWith(String prefix) {
        return find(prefix, false, true);
    }

    private boolean find(String s, boolean isInsert, boolean isPrefix) {
        Node cur = root;

        for (char ch : s.toCharArray()) {
            if (!cur.child.containsKey(ch)) {
                if (isInsert) {
                    cur.child.put(ch, new Node());
                } else {
                    return false;
                }
            }
            cur = cur.child.get(ch);
        }

        if (isInsert)
            cur.count++;
        if (isPrefix)
            return true;

        return cur.count > 0;
    }
}

public class Solution208 {
    public static void main(String[] args) {
        String word = "application";
        String prefix = "app";
        Trie obj = new Trie();
        obj.insert(word);
        boolean param_2 = obj.search(word);
        boolean param_3 = obj.startsWith(prefix);

        System.out.println(param_2);
        System.out.println(param_3);
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */

File: Tree/src/main/python/solution208.py
class Trie:
    def __init__(self):
        self.root = Node()

    def insert(self, word: str) -> None:
        self.find(word, True, False)

    def search(self, word: str) -> bool:
        return self.find(word, False, False)

    def startsWith(self, prefix: str) -> bool:
        return self.find(prefix, False, True)

    def find(self, word, is_insert, is_prefix):
        cur = self.root
        for w in word:
            if not cur.child[ord(w) - ord("a")]:  # type: ignore
                if is_insert:
                    cur.child[ord(w) - ord("a")] = Node()  # type: ignore
                else:
                    return False
            cur = cur.child[ord(w) - ord("a")]  # type: ignore
        if is_insert:
            cur.count += 1  # type: ignore
        if is_prefix:
            return True
        return cur.count > 0  # type: ignore


class Node:
    def __init__(self):
        self.count = 0
        self.child = [None for _ in range(26)]


# Your Trie object will be instantiated and called as such:
word = "abcd"
prefix = "ab"
# word = ""
# prefix = "a"
obj = Trie()
obj.insert(word)
param_2 = obj.search(word)
param_3 = obj.startsWith(prefix)

print(param_3)
